velox/substrait/SubstraitUtils.cpp

```diff
@@ -0,0 +1,160 @@
+      break;
+    default:
+      VELOX_NYI(
+          "Substrait parsing for nullability {} not supported.", nullability);
+  }
+  std::shared_ptr<SubstraitType> substraitType =
+      std::make_shared<SubstraitType>(typeName, nullable);
+  return substraitType;
+}
+
+std::vector<std::shared_ptr<SubstraitParser::SubstraitType>>
+SubstraitParser::parseNamedStruct(const ::substrait::NamedStruct& namedStruct) {
+  const auto& sNames = namedStruct.names();
+  std::vector<std::string> nameList;
+  nameList.reserve(sNames.size());
+  for (const auto& sName : sNames) {
+    nameList.emplace_back(sName);
+  }
+  // Parse Struct.
+  const auto& sStruct = namedStruct.struct_();
+  const auto& sTypes = sStruct.types();
+  std::vector<std::shared_ptr<SubstraitParser::SubstraitType>>
+      substraitTypeList;
+  substraitTypeList.reserve(sTypes.size());
+  for (const auto& type : sTypes) {
+    substraitTypeList.emplace_back(parseType(type));
+  }
+  return substraitTypeList;
+}
+
+int32_t SubstraitParser::parseReferenceSegment(
+    const ::substrait::Expression::ReferenceSegment& sRef) {
+  auto typeCase = sRef.reference_type_case();
+  switch (typeCase) {
+    case ::substrait::Expression::ReferenceSegment::ReferenceTypeCase::
+        kStructField: {
+      auto sField = sRef.struct_field();
+      auto fieldId = sField.field();
+      return fieldId;
+    }
+    default:
+      VELOX_NYI(
+          "Substrait conversion not supported for ReferenceSegment '{}'",
+          typeCase);
+  }
+}
+
+std::vector<std::string> SubstraitParser::makeNames(
+    const std::string& prefix,
+    int size) {
+  std::vector<std::string> names;
+  names.reserve(size);
+  for (int i = 0; i < size; i++) {
+    names.emplace_back(fmt::format("{}_{}", prefix, i));
+  }
+  return names;
+}
+
+std::string SubstraitParser::makeNodeName(int node_id, int col_idx) {
+  return fmt::format("n{}_{}", node_id, col_idx);
+}
+
+std::string SubstraitParser::findSubstraitFunction(
+    const std::unordered_map<uint64_t, std::string>& functionMap,
+    const uint64_t& id) const {
+  if (functionMap.find(id) == functionMap.end()) {
+    VELOX_FAIL("Could not find function id {} in function map.", id);
+  }
+  std::unordered_map<uint64_t, std::string>& map =
+      const_cast<std::unordered_map<uint64_t, std::string>&>(functionMap);
+  return map[id];
+}
+
+std::string SubstraitParser::findVeloxFunction(
+    const std::unordered_map<uint64_t, std::string>& functionMap,
+    const uint64_t& id) const {
+  std::string subFunc = findSubstraitFunction(functionMap, id);
+  std::string veloxFunc = mapToVeloxFunction(subFunc);
+  return veloxFunc;
+}
+
+std::string SubstraitParser::mapToVeloxFunction(
+    const std::string& subFunc) const {
+  if (substraitVeloxFunctionMap.find(subFunc) ==
+      substraitVeloxFunctionMap.end()) {
+    VELOX_FAIL(
+        "Could not find Substrait function {} in function map.", subFunc);
+  }
+  std::unordered_map<std::string, std::string>& map =
```

Comment:
same here.

Reply:
When the function has different names in Substrait and Velox, a mapping relation is needed.
Here we want to map the Substrait key word to Velox's. If not founded, their names are regarded to be the same, so the input should just be returned.